bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,347 @@
+{
+    /// Returns `true` if the given [`Peer`](crate::peer::Peer) is a valid neighbor.
+    fn is_valid(&self, peer: &Peer) -> bool;
+}
+
+// A neighbor is a peer with a distance) metric.
+#[derive(Clone, Debug)]
+pub(crate) struct Neighbor {
+    peer: Peer,
+    distance: Distance,
+}
+
+impl Neighbor {
+    pub(crate) fn new(peer: Peer, distance: Distance) -> Self {
+        Self { peer, distance }
+    }
+
+    pub(crate) fn peer(&self) -> &Peer {
+        &self.peer
+    }
+
+    pub(crate) fn distance(&self) -> Distance {
+        self.distance
+    }
+}
+
+impl fmt::Display for Neighbor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.peer().peer_id(), self.distance())
+    }
+}
+
+impl Eq for Neighbor {}
+impl PartialEq for Neighbor {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+impl PartialOrd for Neighbor {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for Neighbor {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.distance.cmp(&other.distance)
+    }
+}
+
+impl AsRef<Peer> for Neighbor {
+    fn as_ref(&self) -> &Peer {
+        &self.peer
+    }
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct Neighborhood<const N: usize, const INBOUND: bool> {
+    inner: Arc<RwLock<NeighborhoodInner<N, INBOUND>>>,
+}
+
+impl<const N: usize, const INBOUND: bool> Neighborhood<N, INBOUND> {
+    /// Creates a new empty neighborhood.
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Inserts a peer to the neighborhood.
+    pub(crate) fn insert_neighbor(&self, peer: Peer, local: &Local) -> bool {
+        self.write().insert_neighbor(peer, local)
+    }
+
+    /// Removes a peer from the neighborhood.
+    pub(crate) fn remove_neighbor(&self, peer_id: &PeerId) -> Option<Peer> {
+        self.write().remove_neighbor(peer_id)
+    }
+
+    /// Checks whether the candidate is a suitable neighbor.
+    pub(crate) fn is_preferred(&self, candidate: &Neighbor) -> bool {
+        self.write().is_preferred(candidate)
+    }
+
+    /// Picks the first candidate, that is closer than the currently furthest neighbor.
+    pub(crate) fn select_from_candidate_list<'a>(&self, candidates: &'a [&'a Neighbor]) -> Option<&'a Peer> {
+        self.write().select_from_candidate_list(candidates)
+    }
+
+    /// Removes the furthest neighbor from the neighborhood.
+    pub(crate) fn remove_furthest(&self) -> Option<Peer> {
+        self.write().remove_furthest()
+    }
+
+    /// Updates all distances to the neighbors (e.g. after a salt update).
+    pub(crate) fn update_distances(&self, local: &Local) {
+        self.write().update_distances(local);
+    }
+
+    /// Clears the neighborhood, removing all neighbors.
+    pub(crate) fn clear(&self) {
+        self.write().clear();
+    }
+
+    /// Returns the number of neighbors with the neighborhood.
```

Comment:
```suggestion
    /// Returns the number of neighbors in the neighborhood.
```
?

Reply:
I guess I wanted to write "within" :trollface:

Reply:
Within works as well :)

Reply:
Fixed